### ISSUE TITLE
Debug text maybe has a bug

### ISSUE THREAD

**Issue by tathuonglong** [defold/defold#9406]
_Is it a private platform, and a platform specific issue? (i.e. would it break your NDA?)<br>
No, this bug is a bug relate to debug text function, so there is no platform specific issue

**Describe the bug (REQUIRED)**
The position parameter when you use "draw_text" and "draw_text_debug" to debug is always offset from the correct position. "draw_line" works as intented.

**To Reproduce (REQUIRED)**
1. Create a new empty project
2. Go to main.collection, create a new game object inside
3. Create a script, for example "TestBug.script" and add the following lines:
```local color_green = vmath.vector4(0, 1, 0, 1)
local color_red = vmath.vector4(1, 0, 0, 1)
local color_orange = vmath.vector4(1, 0.5, 0, 1)
local color_cyan = vmath.vector4(0, 0.5, 1, 1)

local startRayPoint = vmath.vector3(0, 0, 0)
local endRayPoint = vmath.vector3(570, 450, 0)

function update(self, dt)

	msg.post("@render:", "draw_line", { start_point = startRayPoint, end_point = endRayPoint, color = color_red })
	msg.post("@render:", "draw_text", { text = "ZZZ", position = endRayPoint, color = color_cyan })
end
```
4. Add the TestBug.script to the game object inside main.collection. Press Ctrl+B, and you should see the bug in the screenshot. Here is the screenshot. The "ZZZ" text is offset from the intented position.
![Capture](https://github.com/user-attachments/assets/0415410a-1153-4f64-8f83-b5feb719a9b1)
5. Additionally, you can resize the window and see that the "ZZZ" moving as if its coordinate had been scaled down.

**Expected behavior (REQUIRED)**
The "ZZZ" text should be very close to the end of the line.

**Defold version (REQUIRED):**
  - Version [e.g. 1.9.2], Channel: editor-alpha, Build time: 2024-08-30T12:41:32:202536

**Platforms (REQUIRED):**
 - Platforms: [Windows 10]
 - OS: [Windows 10]
 - Device: [PC]

**Minimal repro case project (OPTIONAL):**
Please attach a minimal project where the bug is reproduced. This will greatly help the person trying to investigate and fix the bug.
[BugTest.zip](https://github.com/user-attachments/files/16939125/BugTest.zip)

**Logs (OPTIONAL):**
Please provide relevant logs from engine, editor or build server. Learn where the logs are stored [here](https://defold.com/manuals/getting-help/#log-files).

**Workaround (OPTIONAL):**
There is no work around

**Screenshots (OPTIONAL):**
I attached a screenshot above.

**Additional context (OPTIONAL):**
None.

**Comment by britzl** [comment]
I tested this on my mac where I have a "retina" screen with display scale of 2.0:

<img width="1072" alt="Screenshot 2024-09-16 at 09 40 17" src="https://github.com/user-attachments/assets/39b43328-c67f-4612-bb4b-05ff3384e75f">

As you can see the "zzz" ends up halfway along the line and not at the end. I can hack a fix like this:

```
function update(self, dt)
	local w,_ = window.get_size()
	local scale = sys.get_config_int("display.width") / w
	msg.post("@render:", "draw_line", { start_point = startRayPoint, end_point = endRayPoint, color = color_red })
	msg.post("@render:", "draw_text", { text = "ZZZ", position = endRayPoint / scale, color = color_cyan })
end
```

<img width="1072" alt="Screenshot 2024-09-16 at 10 05 47" src="https://github.com/user-attachments/assets/61390047-a977-45f6-9078-78f1f833aaa7">

**Comment by britzl** [comment]
@tathuonglong I am guess that you have some display scaling or something set to 175%?

**Comment by koenbollen** [comment]
Yea, I have the similar problem. My Windows display scaling is set to 150%, which causes the same issue.

**Comment by AGulev** [comment]
I can imagine how annoying is sounds, but it's not really a bug, or at least it depends if we consider it's as a bug.

The engine doesn't draw debug lines and text itself. It is specified in render script how it should be drawn.

If you open your `render_script` you'll find that you have a few `render.draw` calls. Some of them draw particles. The other draw sprites, models etc.

But we are interested in two functions:
```
render.draw_debug3d() -- this function draws all the debug lines
render.draw(predicates.debug_text, camera_gui.frustum) -- this function draws debug text
```


and these functions are called in two different blocks with different projections and views:
![CleanShot 2024-09-24 at 17 16 52@2x](https://github.com/user-attachments/assets/d527923f-9882-4458-9a7b-1d5dd3bbadd5)

This is how it looks for me:
![CleanShot 2024-09-24 at 17 19 30@2x](https://github.com/user-attachments/assets/9d2fc598-7750-43d7-bd93-0e4363071fbb)

But if I move `render.draw(predicates.debug_text, ...)`  into the same "block" with the same view and projection:
![CleanShot 2024-09-24 at 17 21 15@2x](https://github.com/user-attachments/assets/0677d132-fd3f-4171-a05a-e0bf3315f258)


So, @britzl we need to decide if it's a bug in `default.render_script` - we just need to fix it, but then debug text will be drawn behind GUI: 
![CleanShot 2024-09-24 at 17 23 06@2x](https://github.com/user-attachments/assets/c9c4c2e5-dd0f-4336-9b54-eba3015d5e07)

or we keep it as is, and everyone can change it using custom render_script

**Comment by britzl** [comment]
> So, @britzl we need to decide if it's a bug in `default.render_script` - we just need to fix it, but then debug text will be drawn behind GUI

A third option would be to draw lines and debug text in a separate pass at the end?

```
function update(self)

    ...

    -- render debug lines and text
    --
    render.set_view(camera_world.view)
    render.set_projection(camera_world.proj)
    render.enable_state(graphics.STATE_BLEND)
    render.draw(predicates.debug_text, camera_gui.frustum)
    render.draw_debug3d()
end
```

**Comment by AGulev** [comment]
> A third option would be to draw lines and debug text in a separate pass at the end?

Maybe then only in debug, to avoid extra set_view/set_projection/enable_state  in release?

**Comment by britzl** [comment]
> > A third option would be to draw lines and debug text in a separate pass at the end?
> 
> Maybe then only in debug, to avoid extra set_view/set_projection/enable_state in release?

Even better!

**Comment by AGulev** [comment]
@tathuonglong @koenbollen sorry, we tested it more, and found that it can't be implemented this way. 
The changes reverted. 

If you need it in you project change your local render_script this way https://github.com/defold/defold/pull/9484/files
It will work for 2d projects, but in general case and 3d projects we have not enough properties in  `draw_text`  message to specify text scale, rotation etc to be able to draw it in world space

**Comment by tathuonglong** [comment]
Thanks guys! I have something to do first and forget about this bug, I have learned to live with it since.
So yeah, this is not a bug, but it should be written down somewhere in the document, so other fellow devs don't have to scratch their head like I did.

**Comment by koenbollen** [comment]
Yes, I figured as much. Already added a ticket to do the same changes to our render script. Thanks for the effort. 👍

**Comment by koenbollen** [comment]
Hey all! I've been trying to implement the same fix in our render_script and I have some more feedback I'd like to share:

In your fix (https://github.com/defold/defold/pull/9484) the debug_text is rendered in world space.
And our project we would like to render the debug text in screen space instead (since it's easier to go from world to screen space then the other way around w/e we want to debug text).

Simplified, this is the relevant render script code:
```lua
-- init:
self.gui_pred = render.predicate({ "gui" })
self.debug_text_pred = render.predicate({ "debug_text" })

-- update:
local window_width = render.get_window_width()
local window_height = render.get_window_height()

... render game ...

local view = IDENTITY
local gui_proj = vmath.matrix4_orthographic(0, window_width, 0, window_height, -1, 1)
render.set_view(view)
render.set_projection(gui_proj)
render.set_viewport(0, 0, window_width, window_height)
render.draw(self.gui_pred, { frustum = gui_proj * view })
if self.is_debug then
  render.draw(self.debug_text_pred, { frustum = gui_proj * view })
end
```
Now, the gui elements (which are just Defold gui stuff) are scaled/rendered at their correct location. But the debug texts are not. The fact that these two builtin Defold system mismatch in their behavior is confusing for me.

On a side note: In the fix, you mix camera_world and camera_gui for the debug rendering.

**Comment by AGulev** [comment]
> If you need it in you project change your local render_script this way https://github.com/defold/defold/pull/9484/files
> It will work for 2d projects, but in general case and 3d projects we have not enough properties in `draw_text` message to specify text scale, rotation etc to be able to draw it in world space

if it's fine for your special case, feel free to to change and use it, but we can't change it in default render script I explained above

**Comment by koenbollen** [comment]
yeah, fair enough. I just wanted to add some more context and not open an new issue.

Let me rephrase one more time, otherwise forget it. 😁 

- If I make a gui label at x=500, y=100 (screen space)
- And a draw_debug_text at x=500, y=100
- When I render those two (the `gui` and `debug_text` predicates) with the same _viewport_, _view_ and _projection_ set
- I would expect them to render at the same position
- They do not (if the display scale != 1)

**Comment by AGulev** [comment]
Yes, it's because of how GUI components works. It takes into account DPI

**Comment by britzl** [comment]
> * When I render those two (the `gui` and `debug_text` predicates) with the same _viewport_, _view_ and _projection_ set
> * I would expect them to render at the same position
> * They do not (if the display scale != 1)

Out of curiosity, what is the gui text node a root node or does it have a parent? What is the adjust mode of the node? I would guess that if Adjust Mode was set to None on the root of the gui component then the text node and debug text would render in the same place, but not otherwise.

**Comment by AGulev** [comment]
> > * When I render those two (the `gui` and `debug_text` predicates) with the same _viewport_, _view_ and _projection_ set
> > * I would expect them to render at the same position
> > * They do not (if the display scale != 1)
> 
> Out of curiosity, what is the gui text node a root node or does it have a parent? What is the adjust mode of the node? I would guess that if Adjust Mode was set to None on the root of the gui component then the text node and debug text would render in the same place, but not otherwise.

Yes, GUI component makes a lot of calculations itself as well.

**Comment by koenbollen** [comment]
> Out of curiosity, what is the gui text node a root node or does it have a parent? What is the adjust mode of the node?

It is a single flat node (text), with Adjust Mode: Fit, the default, we never changed that.

> Yes, GUI component makes a lot of calculations itself as well.

That makes sense, but as a non-Defold-expert, not changing the default GUI behavior: I would expect them to be in the same.

- - -

Again, it's just feedback for the engine, my 50cents. Not asking for fixes/changes just sharing.
